### ISSUE TITLE
taxonomy_vocabulary is no longer an Entity, vocabularies are in config

### ISSUE THREAD

**Issue by bugfolder** [backdrop-contrib/rules#116]
There are many places in Rules that work with the entity type `taxonomy_vocabulary` (as a search on that term will reveal). However, since taxonomy vocabularies are now in config, they're no longer an entity; `entity_get_info()` doesn't include them. So all that code (like `entity_create('taxonomy_vocabulary`, ...)` will no longer work.

Is it possible to modify Rules to handle config-based vocabularies? (or config-based other things?)

**Comment by argiepiano** [comment]
PR #146 

IMPORTANT: For these fixes to work, install the latest dev version Entity Plus AND patch Entity Plus with https://github.com/backdrop-contrib/entity_plus/pull/112

**Comment by argiepiano** [comment]
To test this PR manually you can try the following two rules:

Rules component. Execute and follow directions. This will change the description of a vocabulary.
```json
{ "rules_change_description_of_a_vocabulary" : {
    "LABEL" : "Change description of a vocabulary",
    "PLUGIN" : "action set",
    "OWNER" : "rules",
    "REQUIRES" : [ "rules" ],
    "USES VARIABLES" : {
      "vocabulary" : { "label" : "Vocabulary", "type" : "taxonomy_vocabulary" },
      "new_description" : { "label" : "New description", "type" : "text" }
    },
    "ACTION SET" : [
      { "backdrop_message" : {
          "message" : "I\u0027m about to change the description of vocabulary: [vocabulary:name].",
          "type" : "warning"
        }
      },
      { "data_set" : { "data" : [ "vocabulary:description" ], "value" : [ "new-description" ] } },
      { "vocabulary_save" : { "data" : [ "vocabulary" ], "immediate" : "1" } },
      { "backdrop_message" : { "message" : "The new description is: \u003Cb\u003E[vocabulary:description]\u003C\/b\u003E." } }
    ]
  }
}
```

Rules. This will show an info message when a new term is added to a vocabulary, showing the vocabulary name.
```json
{ "rules_test_voc_1" : {
    "LABEL" : "Test voc 1",
    "PLUGIN" : "reaction rule",
    "OWNER" : "rules",
    "REQUIRES" : [ "rules", "taxonomy" ],
    "ON" : { "taxonomy_term_insert" : [] },
    "DO" : [
      { "vocabulary_fetch" : {
          "USING" : { "id" : [ "term:vocabulary:machine-name" ] },
          "PROVIDE" : { "vocabulary_fetched" : { "vocabulary_fetched" : "Fetched vocabulary" } }
        }
      },
      { "backdrop_message" : { "message" : "You just added a term to vocabulary: [vocabulary-fetched:name]" } }
    ]
  }
}
```

**Comment by bugfolder** [comment]
To test, I created a rule that upon updating a vocabulary, put up a system message with the name of the original and updated vocabulary (possibly in D7). The rule is the same in both D7 and BD:

```
{ "rules_notify_of_updating_a_vocabulary" : {
    "LABEL" : "Notify of updating a vocabulary",
    "PLUGIN" : "reaction rule",
    "OWNER" : "rules",
    "REQUIRES" : [ "rules", "taxonomy" ],
    "ON" : { "taxonomy_vocabulary_update" : [] },
    "DO" : [
      { "backdrop_message" : { "message" : "Vocabulary [vocabulary-unchanged:name] has been updated to [vocabulary:name]." } }
    ]
  }
}
```

But it looks like in BD, the token [vocabulary-unchanged:name] is not supported. Here's the replacement patterns for D7:

![d7](https://user-images.githubusercontent.com/29214084/151285502-03d83e68-b722-4c02-b5a8-f57a66423228.jpg)

And for BD:

![bd](https://user-images.githubusercontent.com/29214084/151285525-b7015a75-c4d4-4e18-a0ec-c676a8161097.jpg)

Is it possible to support the same vocabulary tokens (other than `vid`, which no longer exists)?

**Comment by argiepiano** [comment]
@bugfolder, the issue here is a problem with core, and has nothing to do with Rules of Entity Tokens. Let me explain. 

In D7, vocabularies were entities. This meant that, when you updated or deleted a vocabulary (like any entity), Drupal added the original (pre-saved) entity within the $entity object, under property $entity->original. 

With the switch to non-entity vocabularies, this is not supported in Backdrop. It's not difficult to support, but it has to be done in core, since core is the one that saves or deletes the vocabulary. 

So, we could start an issue in core and hope that it will get some traction. Otherwise, the "vocabulary-unchanged" option is not available to Backdrop Rules.

**Comment by argiepiano** [comment]
I have created a core issue and provided a PR that will provide the "original" vocabulary to the update and presave hooks. This will make it possible to "bring back" the "vocabulary-unchanged"... after this feature is merged in core. 

https://github.com/backdrop/backdrop-issues/issues/5475

In the meantime I can add the code needed for this to Rules, which will become functional as soon as that feature is added. 

Will do so tomorrow.

**Comment by argiepiano** [comment]
I just pushed a commit that adds back a functional implementation of vocabulary_unchanged and all its tokens. It's working great on my end. 

What I did is add code to `rules_taxonomy_vocabulary_presave` to store the original vocabulary under `$vocabulary->original`. This is done only if `original` doesn't exist, meaning that if/when the core PR to add it is merged, it will simply not be added by `rules_taxonomy_vocabulary_presave`.

I tested using the test Rule pasted above in this issue - and also checked the tokens provided. All good. Test and let me know if you see anything fishy.
